Avoid adding master key when credentials already exist

ebihara99999 opened the pull request:
### Summary
Running `rails credentials:edit`,  `MissingKeyError` is raised in the case where `config/master.key` doesn't exist but `config/credentials.yml.enc` already exists.

This commit resolves the following problems:
- the message, the cause of the error is not exact
The real cause is `config/master.key` is missing when `config/credentials.yml.enc` is created firstly.

- another `config/master.key` is added
The created master key is different from the one when `config/credentials.yml.enc` is created firstly, and `MissingKeyError` is railsed when running `rails credentials:edit` again.

#### Reproduction Steps
```
$ bundle exec rails new --dev test_rails_app
$ cd test_rails_app
$ bundle exec rails credentials:edit
New credentials encrypted and saved.
$ mv config/master.key config/master.key.back
$ bundle exec rails credentials:edit

Adding config/master.key to store the master encryption key: c2922c28d9137e2393199a84514fd8c9

Save this in a password manager your team can access.

If you lose the key, no one, including you, can access anything encrypted with it.

      create  config/master.key

Traceback (most recent call last):
	19: from bin/rails:4:in `<main>'
	18: from bin/rails:4:in `require'
	17: from /PATH/TO/src/github.com/rails/rails-dev-box/rails/railties/lib/rails/commands.rb:18:in `<top (required)>'
	16: from /PATH/TO/src/github.com/rails/rails-dev-box/rails/railties/lib/rails/command.rb:46:in `invoke'
	15: from /PATH/TO/src/github.com/rails/rails-dev-box/rails/railties/lib/rails/command/base.rb:65:in `perform'
	14: from /PATH/TO/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/thor-0.20.0/lib/thor.rb:387:in `dispatch'
	13: from /PATH/TO/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/thor-0.20.0/lib/thor/invocation.rb:126:in `invoke_command'
	12: from /PATH/TO/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/thor-0.20.0/lib/thor/command.rb:27:in `run'
	11: from /PATH/TO/src/github.com/rails/rails-dev-box/rails/railties/lib/rails/commands/credentials/credentials_command.rb:26:in `edit'
	10: from /PATH/TO/src/github.com/rails/rails-dev-box/rails/railties/lib/rails/command/helpers/editor.rb:26:in `catch_editing_exceptions'
	 9: from /PATH/TO/src/github.com/rails/rails-dev-box/rails/railties/lib/rails/commands/credentials/credentials_command.rb:27:in `block in edit'
	 8: from /PATH/TO/src/github.com/rails/rails-dev-box/rails/railties/lib/rails/commands/credentials/credentials_command.rb:50:in `change_credentials_in_system_editor'
	 7: from /PATH/TO/src/github.com/rails/rails-dev-box/rails/activesupport/lib/active_support/encrypted_file.rb:54:in `change'
	 6: from /PATH/TO/src/github.com/rails/rails-dev-box/rails/activesupport/lib/active_support/encrypted_configuration.rb:21:in `read'
	 5: from /PATH/TO/src/github.com/rails/rails-dev-box/rails/activesupport/lib/active_support/encrypted_file.rb:42:in `read'
	 4: from /PATH/TO/src/github.com/rails/rails-dev-box/rails/activesupport/lib/active_support/encrypted_file.rb:79:in `decrypt'
	 3: from /PATH/TO/src/github.com/rails/rails-dev-box/rails/activesupport/lib/active_support/messages/rotator.rb:21:in `decrypt_and_verify'
	 2: from /PATH/TO/src/github.com/rails/rails-dev-box/rails/activesupport/lib/active_support/message_encryptor.rb:157:in `decrypt_and_verify'
	 1: from /PATH/TO/src/github.com/rails/rails-dev-box/rails/activesupport/lib/active_support/message_encryptor.rb:183:in `_decrypt'
/PATH/TO/src/github.com/rails/rails-dev-box/rails/activesupport/lib/active_support/message_encryptor.rb:206:in `rescue in _decrypt': ActiveSupport::MessageEncryptor::InvalidMessage (ActiveSupport::MessageEncryptor::InvalidMessage)
```

ebihara99999 committed:
Avoid adding master key when credentials already exist

The case when `config/master.key` doesn't exist but
`config/credentials.yml.enc` already exists, raises MissingKeyError.

This commit resolve the following problems:
- the message, the cause of the error is not exact
- another `config/master.key` is added.

ebihara99999 commented:
Sorry, force push made the the comment and commit order changed. I don't do that when a review is started.

y-yagi commented on railties/test/commands/credentials_test.rb:
```diff
@@ -55,6 +55,15 @@ class Rails::Command::CredentialsCommandTest < ActiveSupport::TestCase
     end
   end
 
+  test "edit_command_does_not_add_master_key_when config/credentials.yml.enc exists" do
```
If using the `test` method, you do not need to use underscores to delimit words. Please use spaces instead.

y-yagi commented on railties/lib/rails/commands/credentials/credentials_command.rb:
```diff
@@ -20,6 +20,13 @@ def edit
 
         ensure_editor_available(command: "bin/rails credentials:edit") || (return)
+
+        if File.exist?("config/credentials.yml.enc") && !File.exist?("config/master.key")
```
This check does not work properly when `ENV["RAILS_MASTER_KEY"]` is used. 
When `ENV["RAILS_MASTER_KEY"]` is used, there is no `config/master.key` but data can be decrypted.

ebihara99999 replied:
Thanks, I will add the condition.

ebihara99999 committed:
Use speces instead of underscores

ebihara99999 committed:
Add master key env condition

ebihara99999 commented:
@y-yagi 
Would you review again when you have time?

y-yagi commented on railties/lib/rails/commands/credentials/credentials_command.rb:
```diff
@@ -20,6 +20,13 @@ def edit
 
         ensure_editor_available(command: "bin/rails credentials:edit") || (return)
+
+        if File.exist?("config/credentials.yml.enc") && !master_key_defined?
```
I do not want to use filename directly (Because if we want to change the filename, there are more places to change). 
Since this path can be obtained from `EncryptedFile#content_path`, can use it as follows..

```diff 
-        ensure_master_key_has_been_added if Rails.application.credentials.key.nil?
+
+        if Rails.application.credentials.key.nil?
+          return say missing_credentials_message if Rails.application.credentials.content_path.exist?
+
+          ensure_master_key_has_been_added
+        end
```

y-yagi replied:
An error message to be showed when the same problem occurs with the `credentials:show` command already exists. 
https://github.com/rails/rails/blob/a9692941600c34afbc8d60d15cb5f179d30f67e5/railties/lib/rails/commands/credentials/credentials_command.rb#L72 

What do you think about reusing it?

ebihara99999 replied:
I totally agree with using it. Thanks for the suggestions including the previous.

ebihara99999 committed:
Use `Rails.application.credentials.key` instead of `master_key_defined?`

ebihara99999 committed:
Change not using a path

ebihara99999 committed:
Reuse `.missing_credentials_message`

y-yagi commented on railties/lib/rails/commands/credentials/credentials_command.rb:
```diff
@@ -20,6 +20,11 @@ def edit
         ensure_editor_available(command: "bin/rails credentials:edit") || (return)
+
+        if Rails.application.credentials.key.nil?
+          return say missing_credentials_message if Rails.application.credentials.content_path.exist?
```
Can merge condition?

```diff 
+        if Rails.application.credentials.key.nil? && Rails.application.credentials.content_path.exist?
+          return say missing_credentials_message
+        end
```

y-yagi commented:
I made a minor comment but the others are LGTM.   
When finished reacting to that comment, please squash your commits into a single commit.

kaspth commented on railties/lib/rails/commands/credentials/credentials_command.rb:
```diff
@@ -52,7 +57,6 @@ def change_credentials_in_system_editor
           end
         end
 
-
```
Separation of method concerns. Just leave it in, thanks.

kaspth replied:
Still needs to be left in.

ebihara99999 replied:
What do you think is needed? Seperate conditions in `#ensure_master_key_has_been_added` out here?

kaspth commented on railties/lib/rails/commands/credentials/credentials_command.rb:
```diff
@@ -20,6 +20,11 @@ def edit
+
+        if Rails.application.credentials.key.nil?
+          return say missing_credentials_message if Rails.application.credentials.content_path.exist?
+        end
```
This breaks the flow of the command. Before we had 4 lines of `ensure_*` methods but now there's something here at a more granular level of abstraction.

I don't think `missing_credentials_message` makes sense here either. It's `else` branch mentions using `credentials:edit`, which is what the user is already running. Confusing usage.

Aren't both checks just variants of `ensure_master_key_has_been_added`? If we haven't found a key we should add it but only if we're generating new credentials?

ebihara99999 replied:
@kaspth

Thanks for the comment! The way of abstraction sounds reasonable for me, but I'm not completely sure what you meant.

You prefer to prepare another message but `missing_credentials_message` and abstract the condition `Rails.application.credentials.key.nil? && !Rails.application.credentials.content_path.exist?` as follows:

```ruby
def edit
  .... 
  return say "some messages" unless master_key_should_be_added?
  ....
end

private
  def master_key_should_be_added?
    Rails.application.credentials.key.nil? && !Rails.application.credentials.content_path.exist?
  end
```

kaspth replied:
I'm not convinced we need another message. What do we need to explain to the user here? That we aren't doing anything?

I was alluding to just doing this:

```ruby
def ensure_master_key_has_been_added
  if Rails.application.credentials.key.nil? && !Rails.application.credentials.content_path.exist?
    master_key_generator.add_master_key_file
    master_key_generator.ignore_master_key_file
  end
end
```

kaspth reviewed (commented):
Thanks for the pull request! This is too hard to understand code-wise as is, I've left a question or two for you.

ebihara99999 committed:
Do not break the flow of the edit command

Move the condition of the existence of a credentials file into
`.ensure_master_key_has_been_added`.

ebihara99999 committed:
Merge remote-tracking branch 'origin/master' into avoid-creating-master-key

ebihara99999 commented:
@kaspth 
Would you review again?
The last commit is to take in https://github.com/rails/rails/pull/33566 to pass the CI.

kaspth commented on railties/test/commands/credentials_test.rb:
```diff
@@ -76,7 +86,7 @@ class Rails::Command::CredentialsCommandTest < ActiveSupport::TestCase
     def run_edit_command(editor: "cat")
       switch_env("EDITOR", editor) do
-        rails "credentials:edit"
+        rails "credentials:edit", allow_failure: true
```
Too lenient. Only one instance of the command is expected to fail. Let's pass it in as a keyword argument from your test.

kaspth commented on railties/lib/rails/commands/credentials/credentials_command.rb:
```diff
@@ -38,8 +38,13 @@ def show
+            master_key_generator.add_master_key_file
+            master_key_generator.ignore_master_key_file
+          else
+            say missing_credentials_message
```
So we're saying credentials are missing if the content path exists? That doesn't make sense to me.

Why won't my earlier suggestion do?

```ruby
def ensure_master_key_has_been_added
  if Rails.application.credentials.key.nil? && !Rails.application.credentials.content_path.exist?
    master_key_generator.add_master_key_file
    master_key_generator.ignore_master_key_file
  end
end
```

ebihara99999 replied:
@kaspth 
There are the 2 differences from yours

1. `Rails.application.credentials.key.nil?` is omitted  
because this is already checked when `ensure_master_key_has_been_added` is called.

2. exits and show `missing_credentials_message`  if it steps into the else branch

Without exit, if you run `rails credentials:edit`,  you edit `/tmp/2293.credentials.yml` like the following screen shot.

<img width="895" alt="2018-08-12 12 26 39" src="https://user-images.githubusercontent.com/7948153/43998206-27277966-9e2b-11e8-9caa-cb41b1b59db9.png">

 If nothing changes in it, the message `New credentials encrypted and saved.` shows but `credentials.yml.enc` is not saved, which I check the time stamp. But adding some changes and save it,  the error below occurs:
```
$ EDITOR=vi bundle exec rails credentials:edit
Traceback (most recent call last):
	19: from bin/rails:4:in `<main>'
	18: from bin/rails:4:in `require'
	17: from /vagrant/rails/railties/lib/rails/commands.rb:18:in `<top (required)>'
	16: from /vagrant/rails/railties/lib/rails/command.rb:46:in `invoke'
	15: from /vagrant/rails/railties/lib/rails/command/base.rb:65:in `perform'
	14: from /var/lib/gems/2.5.0/gems/thor-0.20.0/lib/thor.rb:387:in `dispatch'
	13: from /var/lib/gems/2.5.0/gems/thor-0.20.0/lib/thor/invocation.rb:126:in `invoke_command'
	12: from /var/lib/gems/2.5.0/gems/thor-0.20.0/lib/thor/command.rb:27:in `run'
	11: from /vagrant/rails/railties/lib/rails/commands/credentials/credentials_command.rb:26:in `edit'
	10: from /vagrant/rails/railties/lib/rails/command/helpers/editor.rb:26:in `catch_editing_exceptions'
	 9: from /vagrant/rails/railties/lib/rails/commands/credentials/credentials_command.rb:27:in `block in edit'
	 8: from /vagrant/rails/railties/lib/rails/commands/credentials/credentials_command.rb:55:in `change_credentials_in_system_editor'
	 7: from /vagrant/rails/activesupport/lib/active_support/encrypted_file.rb:54:in `change'
	 6: from /vagrant/rails/activesupport/lib/active_support/encrypted_file.rb:68:in `writing'
	 5: from /vagrant/rails/activesupport/lib/active_support/encrypted_configuration.rb:29:in `write'
	 4: from /vagrant/rails/activesupport/lib/active_support/encrypted_file.rb:49:in `write'
	 3: from /vagrant/rails/activesupport/lib/active_support/encrypted_file.rb:75:in `encrypt'
	 2: from /vagrant/rails/activesupport/lib/active_support/message_encryptor.rb:151:in `encrypt_and_sign'
	 1: from /vagrant/rails/activesupport/lib/active_support/message_encryptor.rb:169:in `_encrypt'
/vagrant/rails/activesupport/lib/active_support/message_encryptor.rb:169:in `key=': key must be 16 bytes (ArgumentError)
```

That's thy I implement like this. I think `exit` is needed and then some messages are necessary.
But if you don't like to use `missing_credentials_message` to show the message, I will modify it. I would like a suggestion what the message is.

kaspth replied:
Doesn't make sense to me that there's a conditional both within and outside ensure_master_key_has_been_added. They should both go in there.

Sure, but we've already got the line `say Rails.application.credentials.read.presence || missing_credentials_message`, won't this be in a state where the credentials can't be read?

kaspth reviewed (commented):
Another round poured! 🍺

Please squash your commits down to 1 as well.

ebihara99999 committed:
Add allow_failure as an option

ebihara99999 commented:
:memo: 
`environment = 'staging'`にするとL13とL21のテストが落ちるし、`environment = 'production'`とすると1つ目のテストが落ちる（ローカル）

ebihara99999 committed:
Include a condition into `ensure_master_key_has_been_added`

oniofchaos commented:
👍 after fixing merge conflicts

ebihara99999 commented:
@oniofchaos 
Thanks! I'll fix the conflicts soon.

ebihara99999 commented:
@y-yagi @kaspth @oniofchaos 

I'm closing the PR. 

The problem trying to fix here seems to have been fixed in #33521.
Following the same reproduction steps, the problem never occurs again.

Thanks for all reviews!!

```
vagrant@rails-dev-box:/vagrant/rails/test_app$ ls config/master.key config/credentials.yml.enc
ls: cannot access 'config/master.key': No such file or directory
ls: cannot access 'config/credentials.yml.enc': No such file or directory
vagrant@rails-dev-box:/vagrant/rails/test_app$ EDITOR=vi bin/rails credentials:edit
Adding config/master.key to store the master encryption key: 81a342abcaae8f5e1118e82fea292688

Save this in a password manager your team can access.

If you lose the key, no one, including you, can access anything encrypted with it.

      create  config/master.key

New credentials encrypted and saved.
vagrant@rails-dev-box:/vagrant/rails/test_app$ ls config/master.key config/credentials.yml.enc
config/credentials.yml.enc  config/master.key
vagrant@rails-dev-box:/vagrant/rails/test_app$ mv config/master.key config/master.key.back
vagrant@rails-dev-box:/vagrant/rails/test_app$ EDITOR=vi bin/rails credentials:edit
Missing master key to decrypt credentials. See `rails credentials:help`
vagrant@rails-dev-box:/vagrant/rails/test_app$ ls config/master.key config/credentials.yml.enc
ls: cannot access 'config/master.key': No such file or directory
config/credentials.yml.enc
vagrant@rails-dev-box:/vagrant/rails/test_app$
```

kaspth commented:
Nice!
